jetcd-core/src/test/java/io/etcd/jetcd/KVNamespaceTest.java

```diff
@@ -0,0 +1,500 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.etcd.jetcd;
+
+import com.google.common.base.Charsets;
+import com.google.protobuf.ByteString;
+import io.etcd.jetcd.kv.DeleteResponse;
+import io.etcd.jetcd.kv.GetResponse;
+import io.etcd.jetcd.kv.PutResponse;
+import io.etcd.jetcd.kv.TxnResponse;
+import io.etcd.jetcd.launcher.EtcdCluster;
+import io.etcd.jetcd.launcher.EtcdClusterFactory;
+import io.etcd.jetcd.op.Cmp;
+import io.etcd.jetcd.op.CmpTarget;
+import io.etcd.jetcd.op.Op;
+import io.etcd.jetcd.options.DeleteOption;
+import io.etcd.jetcd.options.GetOption;
+import io.etcd.jetcd.options.PutOption;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Test cases focusing on using KVClient and TxnClient with non-empty namespace.
+ */
+public class KVNamespaceTest {
+  private static final ByteSequence END_KEY = ByteSequence.from(new byte[] {0});
+
+  public static final EtcdCluster CLUSTER = EtcdClusterFactory.buildCluster(
+      "etcd-kv-namespace-test", 3 ,false);
+  private KV kvClient;
+  private KV kvClientWithNamespace;
+  private KV kvClientWithNamespace2;
+
+  private static Integer keyIndex = -1;
+
+  @BeforeClass
+  public static void setUp() {
+    CLUSTER.start();
+  }
+
+  @AfterClass
+  public static void tearDown() throws IOException {
+    CLUSTER.close();
+  }
+
+  @After
+  public void cleanUpCase() {
+    if (kvClient != null) {
+      kvClient.close();
+      kvClient = null;
+    }
+    if (kvClientWithNamespace != null) {
+      kvClientWithNamespace.close();
+      kvClientWithNamespace = null;
+    }
+    if (kvClientWithNamespace2 != null) {
+      kvClientWithNamespace2.close();
+      kvClientWithNamespace2 = null;
+    }
+  }
+
+  @Test
+  public void testPrefixNamespace() {
+    class TestCase {
+      public ByteSequence namespace;
+      public ByteString key;
+      public ByteString end;
+      public ByteString wKey;
+      public ByteString wEnd;
+
+      public TestCase(ByteSequence namespace, ByteString key, ByteString end,
+          ByteString wKey, ByteString wEnd) {
+        this.namespace = namespace;
+        this.key = key;
+        this.end = end;
+        this.wKey = wKey;
+        this.wEnd = wEnd;
+      }
+    }
+    List<TestCase> testCases = Arrays.asList(
```

Comment:
I know you can do similar testing by different cases using dataprovider but i don't want to testNG in jetcd and actually want to get rid of it. as you can see jetcd testing has a mix of junit and testNG. it gets confusing. Is it possible to do the same thing using junit's Parameterized tests here?

Reply:
As to parameterized tests, there is a shortcoming of parameterized test in junit4:
we can only do parameterize test per test class. That is to say, if we configure two sets of parameters, then all the cases in this test class will be executed twice. While there should be some situations where not all the tests need to be parameterized, furthermore maybe not all the cases use the same parameter source.

A possible solution is to divide cases into test classes by their parameter sets, which might lose readability of the integral test design of a module.

In junit5 parameter source is introduced, while testNG also provides similar api. Since we decided to remove testNG, I wonder if we could change junit from junit4 to junit5 ?

Ref:
https://blog.parasoft.com/how-to-create-junit-parameterized-tests-faster
https://github.com/junit-team/junit4/wiki/Parameterized-tests
https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests

Reply:
+1 for junit 5

Reply:
Fine with me to use junit 5